pangeo_forge_recipes/recipes/xarray_zarr.py

```diff
@@ -206,303 +682,178 @@ def _set_target_chunks(self):
-        input_positions = [self.input_position(input_key) for input_key in input_keys]
-        start = chunk_bounds[min(input_positions)]
-        stop = chunk_bounds[max(input_positions) + 1]
-
-        this_chunk_conflicts = set()
-        for k in input_keys:
-            # for multi-variable recipes, the confilcts will usually be the same
-            # for each variable. using a set avoids duplicate locks
-            for input_conflict in all_chunk_conflicts[self.input_position(k)]:
-                this_chunk_conflicts.add(input_conflict)
-        region_slice = slice(start, stop)
-        return {self._concat_dim: region_slice}, this_chunk_conflicts
+        return Delayed(key, dsk)
 
-    def iter_chunks(self):
-        for k in self._chunks_inputs:
-            yield k
+    def to_prefect(self):
+        """Compile the recipe to a Prefect.Flow object."""
+        from prefect import Flow, task, unmapped
 
-    def get_input_meta(self, *input_keys: Sequence[InputKey]) -> Dict:
-        # getitems should be async; much faster than serial calls
-        if self.metadata_cache is None:
-            raise ValueError("metadata_cache is not set.")
-        return self.metadata_cache.getitems([_input_metadata_fname(k) for k in input_keys])
-
-    def input_position(self, input_key):
-        # returns the index position of an input key wrt the concat_dim
-        concat_dim_axis = list(self.file_pattern.dims).index(self._concat_dim)
-        return input_key[concat_dim_axis]
-
-    def calculate_sequence_lens(self):
-        if self._nitems_per_input:
-            return list((self._nitems_per_input,) * self.file_pattern.dims[self._concat_dim])
-
-        # read per-input metadata; this is distinct from global metadata
-        # get the sequence length of every file
-        # this line could become problematic for large (> 10_000) lists of files
-        input_meta = self.get_input_meta(*self._inputs_chunks)
-        # use a numpy array to allow reshaping
-        all_lens = np.array([m["dims"][self._concat_dim] for m in input_meta.values()])
-        all_lens.shape = list(self.file_pattern.dims.values())
-        # check that all lens are the same along the concat dim
-        concat_dim_axis = list(self.file_pattern.dims).index(self._concat_dim)
-        selector = [slice(0, 1)] * len(self.file_pattern.dims)
-        selector[concat_dim_axis] = slice(None)  # this should broadcast correctly agains all_lens
-        sequence_lens = all_lens[tuple(selector)]
-        if not (all_lens == sequence_lens).all():
-            raise ValueError(f"Inconsistent sequence lengths found: f{all_lens}")
-        return sequence_lens.squeeze().tolist()
+        cache_input_task = task(self._cache_input, name="cache_input")
+        prepare_target_task = task(self._prepare_target, name="prepare_target")
+        store_chunk_task = task(self._store_chunk, name="store_chunk")
+        finalize_target_task = task(self._finalize_target, name="finalize_target")
```

Comment:
Here you not using `_finalize_target` rather than `finalize_target`, etc.

Is there any point in maintaining the `finalize_target` methods if we are going to remove the existing `to_pipelines` method in `BaseRecipe`?

Reply:
I think having the top-level property `.finalize_target` is still really nice for users who are developing / debugging recipes. We want to keep that, while still allowing Dask and Prefect to access the (partially applied) function itself, so they can wrap it in a Task.